File: 416. Partition Equal Subset Sum.py
'''
https://leetcode.com/problems/partition-equal-subset-sum/
Given a non-empty array nums containing only positive integers, 
find if the array can be partitioned into two subsets such that the sum of elements in both subsets is equal.

Example 1:
Input: nums = [1,5,11,5]
Output: true
Explanation: The array can be partitioned as [1, 5, 5] and [11].
'''
'''
Time:O(n*m)
Space:O(n*m)
'''
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        if sum(nums)%2==1:  return False
        tar = sum(nums)//2
        dp = [[False for _ in range(tar+1)]for _ in range(len(nums)+1)]
        dp[0][0] = True
        for i in range(1, len(nums)+1):
            for j in range(1, tar+1):
                cur = nums[i-1]
                if cur>j:
                    dp[i][j] = dp[i-1][j]
                else:
                    dp[i][j] = dp[i-1][j] or dp[i-1][j-cur]
        return dp[-1][-1]
        

File: 416. Partition Equal Subset Sum.java
/*
https://leetcode.com/problems/partition-equal-subset-sum/
Given a non-empty array nums containing only positive integers, 
find if the array can be partitioned into two subsets such that the sum of elements in both subsets is equal.

Example 1:
Input: nums = [1,5,11,5]
Output: true
Explanation: The array can be partitioned as [1, 5, 5] and [11].
*/
/*
Time:O(n*m)
Space:O(n)
*/
class Solution {
    public boolean canPartition(int[] nums) {
        int sum = IntStream.of(nums).sum();
        int subsum = sum/2;
        if (sum%2 != 0)  return false;
        boolean[] dp = new boolean[subsum+1];
        dp[0] = true;
        for (int num:nums){
            for (int j=subsum; j>=num; j--)
                dp[j] = dp[j]|dp[j-num];
        }
        return dp[subsum];
    }
}

/*
Time:O(n*m)
Space:O(n*m)
*/class Solution {
    public boolean canPartition(int[] nums) {
        int sum = IntStream.of(nums).sum();
        int subsum = sum/2;
        if (sum%2 != 0)  return false;
        boolean[][] dp = new boolean[nums.length+1][subsum+1];
        dp[0][0] = true;
        for (int i=1; i<nums.length+1; i++){
            int cur = nums[i-1];
            for (int j=0; j<subsum+1; j++){
                if (j<cur)  dp[i][j] = dp[i-1][j];
                else    dp[i][j] = (dp[i-1][j] || dp[i-1][j-cur]);
            }
        }
        return dp[nums.length][subsum];
    }
}
